### DISCUSSION TITLE
Bugs in SSD1306 example

### DISCUSSION THREAD

**Discussion by marklacas** [tabemann/zeptoforth#30] [Q&A]
I've been trying to get the SSD1306 example to work but failing.  Everything loads ok with no errors.  When I run it with the $ff in the last chunk of code it looks like the first image.  If I change th $ff to $00 it looks like the second image.  I've tried messing about with the rectangles and such but not having any success.  Any thoughts?
![IMG_0094](https://user-images.githubusercontent.com/566089/221320406-6f07de25-eb2f-41d7-b98d-eb3c198fe6c9.jpeg)
![IMG_0095](https://user-images.githubusercontent.com/566089/221320435-aee1512c-11c7-4108-af42-9c3266084654.jpeg)

**Comment by tabemann** [comment]
That is puzzling, because it implies that `clear-bitmap` is not being called on the framebuffer's backing bitmap, and hence you are seeing garbage that you are XORing with your rectangle drawing operations, but I just checked, and the constructor for a `<ssd1306>` object calls the `<bitmap>` constructor, which calls `clear-bitmap`. Could I see the exact code you are working with, because I have more than one SSD1306 example in the `test/` directory?

**Reply by marklacas** [reply]
#include bitmap.fs
#include bitmap_utils.fs
#include font.fs
#include simple_font_big.fs
#include ssd1306.fs

oo import
bitmap import
ssd1306 import
font import
simple-font import

128 constant my-width
64 constant my-height

my-width my-height bitmap-buf-size constant my-buf-size
my-buf-size 4 align buffer: my-buf
<ssd1306> class-size buffer: my-ssd1306
14 15 my-buf my-width my-height SSD1306_I2C_ADDR 1 <ssd1306> my-ssd1306 init-object

init-simple-font

s" Hello, World!" my-width 2 / 7 13 * 2 / - my-height 2 / 8 2 / - my-ssd1306 a-simple-font set-string
$00 my-width 4 / my-width 2 / 0 my-height my-ssd1306 xor-rect-const
my-ssd1306 update-display

**Reply by tabemann** [reply]
I notice you have loaded a source file named `simple_font_big.fs`. Am I right in assuming that this is a new font you have created? Could you post the source of this font or link to it somewhere, as this might explain what you are seeing.

**Reply by marklacas** [reply]
Same font file as yours.  My plan was to make a bigger font when I get this going.  But I have made no changes in it yet.

**Reply by tabemann** [reply]
When I copy-and-pasted your code into zeptocom.js, it caused a HardFault. It turns out that in your code above two instances of `<ssd1306>` were missing, probably eaten by GitHub's Markdown parser at some point along the way. However, this is probably not the issue you saw, because if this was the case with your code it would have never displayed what you saw in the first place; it would have simply crashed instead.

I notice that your screen dimensions look thinner in the Y dimension than mine. My screen is 128x64, but there are also 128x32 screens out there as well. Could you perhaps have acquired a 128x32 screen? This might explain what you are seeing, because there are definite issues with the image that do not seem to imply any particular intrinsic software bug. If you want to use a 128x32 screen, replace `64 constant my-height` with `32 constant my-height` in your code here.

As for the missing `<ssd1306>`s, here is the same code with fixes made:

```
#include bitmap.fs
#include bitmap_utils.fs
#include font.fs
#include simple_font.fs
#include ssd1306.fs

oo import
bitmap import
ssd1306 import
font import
simple-font import

128 constant my-width
64 constant my-height

my-width my-height bitmap-buf-size constant my-buf-size
my-buf-size 4 align buffer: my-buf
<ssd1306> class-size buffer: my-ssd1306
14 15 my-buf my-width my-height SSD1306_I2C_ADDR 1 <ssd1306> my-ssd1306 init-object

init-simple-font

s" Hello, World!" my-width 2 / 7 13 * 2 / - my-height 2 / 8 2 / - my-ssd1306 a-simple-font set-string
$00 my-width 4 / my-width 2 / 0 my-height my-ssd1306 xor-rect-const
my-ssd1306 update-display
```

The code worked for me after I made those fixes; my SSD1306-based display currently shows "Hello, World!".

![20230227_203701](https://user-images.githubusercontent.com/4637032/221740121-8aa55bc0-77a4-42e3-88dc-171441de3243.jpg)

Here is the same, but with `$00` replaced with `$FF`:

![20230227_205152](https://user-images.githubusercontent.com/4637032/221741905-5c5fb9ed-fcde-4425-96fe-abdffafc5e12.jpg)

(Forgive the ugly way the image came out on my phone camera - this is an artifact of how the SSD1306-based screen draws itself.)

When pasting code, remember to either indent it by four characters, place it between three backticks before and after, or, if within a line, between single backticks.

**Reply by marklacas** [reply]
I bought the same display as you have and it works great.  Still can't get the small display to show properly.
![IMG_0102](https://user-images.githubusercontent.com/566089/222512243-af7269c4-5efe-45b2-b6f8-2cdd0a5639a6.jpeg)
![IMG_0103](https://user-images.githubusercontent.com/566089/222512272-3b184938-3a3d-4774-b175-f982ea673a8b.jpeg)

**Reply by marklacas** [reply]
I did change the height from 64 to 32.  Didn't really make a difference.

**Reply by tabemann** [reply]
Leave the height at 32, that is correct. However, I think there is another change you need to make, to `ssd1306.fs`:

Starting at line 244, replace the following:

```
      \ Set COM pins to alternative COM pin mode
      SSD1306_SETCOMPINS >cmd $12 >cmd
```

with:

```
      \ Set COM pins mode
      SSD1306_SETCOMPINS >cmd self bitmap-rows @ 32 > if $12 else $02 then >cmd
```

If my thinking is correct, this should work for both 128x64 and 128x32 screens. I have tested it on my 128x64 screen and it works for me.

Let me know if this works for you - if it does, I will update my code in the repository accordingly.

**Reply by marklacas** [reply]
Well, that worked perfectly!  Great job!
Thanks for being so responsive and taking care of this.
Another reason I love ZeptoForth.
BTW - A friend updated Zeptocom so it works from a local file system.  No web server.  I use it every day this way.
I can provide the code if you want.

**Reply by tabemann** [reply]
Thanks for trying out the SSD1306 driver on both 128x64 and 128x32 displays and testing my change to it! I am glad it worked for you. I will incorporate the changes into zeptoforth now; they will be in the `devel` branch today, and I will merge them along with a host of other changes into the `master` branch once something else I am currently working on (the ESP8285 WiFi radio driver for the Wio RP2040) is ready (or I finally decide that the ESP8285 WiFi radio is simply too tricky to work with and give up on it).

About zeptocom.js, I would highly suggest you start a pull request on the [zeptocom.js GitHub](https://github.com/tabemann/zeptocomjs/) to incorporate your friend's changes to zeptocom.js. Give your friend my thanks for working on it.
